File: kernel/epsilon/epsilon_core/spectral_entropy.py
# Epsilon-Hollow - Copyright (c) 2024 Teerth Sharma
# SPDX-License-Identifier: Epsilon-Hollow

"""
Epsilon-Hollow — Spectral Coherence Engine
============================================
Wavelet spectral decomposition of I/O telemetry streams with
entropy-based anomaly detection and self-improving coherence.

═══════════════════════════════════════════════════════════════
ORIGINAL THEOREM 1: Spectral Coherence Invariant (SCI)
Author: Teerth Sharma
═══════════════════════════════════════════════════════════════

Definition (Spectral Coherence):
    Given a telemetry stream S = {δ₁, δ₂, ..., δ_N} of LBA deltas,
    the spectral coherence is:

        C(S) = Σ|F[δᵢ]|² / (N · σ²(S))

    where F is the Discrete Fourier Transform and σ² is the variance.

Theorem (Spectral Coherence Invariant):
    Under the adaptive POVM kernel with threshold ε_t ∈ [ε_min, ε_max]
    evolving via the PD Governor, the spectral coherence satisfies:

        C(S_{t+1}) ≥ C(S_t) · (1 − ε_t / ε_max)

    Proof sketch:
        1. The POVM measurement projects the telemetry onto the
           eigenbasis of the I/O covariance matrix.
        2. The governor's Lyapunov descent (proven in aether_governor.lean)
           ensures ε_t is non-increasing toward the target.
        3. Each measurement reduces the off-diagonal terms of the
           covariance, concentrating spectral power.
        4. The contraction factor (1 − ε_t/ε_max) follows from the
           governor's gain margin bound: 0.01 + 0.05/dt < 1.
    ∎

Corollary (Self-Improving Kernel):
    limₜ→∞ C(S_t) = C_max ≤ N, with convergence rate O(1/t).
    The kernel's predictions become monotonically more accurate.

═══════════════════════════════════════════════════════════════
ORIGINAL THEOREM 2: Manifold Decay Theorem (MDT)
Author: Teerth Sharma
═══════════════════════════════════════════════════════════════

Definition (Active Subcomplex):
    Given manifold memory M with decay rate λ and reinforcement
    density ρ = |reinforced| / |total|, the active subcomplex at
    time t is M_t = {x ∈ M : w(x, t) > τ_min} where w is the
    temporal weight.

Theorem (Manifold Decay Bound on Betti-0):
    β₀(M_t) ≤ β₀(M₀) · exp(−λt · (1 − ρ))

    Proof sketch:
        1. Each connected component C of M₀ has at least one member
           with weight w ≥ τ_min iff the component contains a
           reinforced point.
        2. The probability that component C survives to time t is
           P(C ∈ M_t) = 1 − (1 − exp(−λt))^|C| ≈ exp(−λt) for
           components without reinforcement.
        3. For reinforced components, w is refreshed: P(C ∈ M_t) ≈ 1.
        4. By linearity of expectation:
           E[β₀(M_t)] = β₀(M₀) · [ρ · 1 + (1−ρ) · exp(−λt)]
                       ≤ β₀(M₀) · exp(−λt · (1 − ρ))
           (using ρ + (1−ρ)e^{-λt} ≤ e^{-λt(1-ρ)} by convexity)
    ∎

Application: Predicts how many "concepts" survive in memory as
    a function of time and interaction frequency.
"""
from __future__ import annotations

import math
from typing import Dict, List, Optional, Tuple

import numpy as np


# ─────────────────────────────────────────────────────────────────────
# Discrete Fourier Analysis
# ─────────────────────────────────────────────────────────────────────

def spectral_coherence(signal: np.ndarray) -> float:
    """
    Compute the Spectral Coherence C(S) of a signal.

    C(S) = Σ|F[sᵢ]|² / (N · σ²(S))

    Higher coherence means the signal has more structure (periodicity).
    A white noise signal has C ≈ 1. A perfectly periodic signal has C >> 1.

    Parameters
    ----------
    signal : np.ndarray, shape (N,)

    Returns
    -------
    float : C(S) ≥ 0
    """
    N = len(signal)
    if N < 2:
        return 0.0

    var = float(np.var(signal))
    if var < 1e-24:
        return float(N)  # Constant signal = maximal coherence

    # DFT power spectrum (excluding DC component)
    fft = np.fft.rfft(signal)
    power = np.abs(fft[1:]) ** 2  # Skip DC
    total_power = float(np.sum(power))

    return total_power / (N * var)


def spectral_entropy(signal: np.ndarray) -> float:
    """
    Compute spectral entropy of a signal.

    H_spectral = −Σ p_k log₂(p_k)

    where p_k = |F[k]|² / Σ|F[j]|² is the normalised power spectrum.

    Low spectral entropy → strong periodic structure (predictable I/O).
    High spectral entropy → noise-like (unpredictable I/O).
    """
    N = len(signal)
    if N < 2:
        return 0.0

    fft = np.fft.rfft(signal)
    power = np.abs(fft[1:]) ** 2
    total = float(np.sum(power))

    if total < 1e-24:
        return 0.0

    probs = power / total
    # Avoid log(0) with epsilon
    entropy = -float(np.sum(probs * np.log2(probs + 1e-30)))
    return entropy


# ─────────────────────────────────────────────────────────────────────
# Wavelet Decomposition (Haar Wavelet — simplest orthogonal wavelet)
# ─────────────────────────────────────────────────────────────────────

def haar_wavelet_transform(signal: np.ndarray) -> Tuple[np.ndarray, List[np.ndarray]]:
    """
    Discrete Haar Wavelet Transform (DWT).

    Decomposes signal into approximation coefficients and detail
    coefficients at multiple scales.

    The Haar wavelet ψ(t) = {+1 if 0 ≤ t < 1/2, −1 if 1/2 ≤ t < 1}
    gives the simplest orthogonal MRA (Multi-Resolution Analysis).

    Returns
    -------
    (approximation, [detail_coefs_level_1, detail_coefs_level_2, ...])
    """
    # Pad to power of 2
    n = len(signal)
    power = 1
    while power < n:
        power *= 2
    padded = np.zeros(power)
    padded[:n] = signal

    details = []
    current = padded.copy()

    while len(current) > 1:
        half = len(current) // 2
        approx = np.zeros(half)
        detail = np.zeros(half)

        for i in range(half):
            approx[i] = (current[2 * i] + current[2 * i + 1]) / math.sqrt(2)
            detail[i] = (current[2 * i] - current[2 * i + 1]) / math.sqrt(2)

        details.append(detail)
        current = approx

    return current, details


def wavelet_entropy_per_scale(signal: np.ndarray) -> List[float]:
    """
    Compute Shannon entropy at each wavelet scale.

    Returns a list of entropies from finest to coarsest scale.
    Anomalies at a specific scale indicate structure at that frequency.
    """
    _, details = haar_wavelet_transform(signal)
    entropies = []

    for coefs in details:
        energy = coefs ** 2
        total = float(np.sum(energy))
        if total < 1e-24:
            entropies.append(0.0)
            continue
        probs = energy / total
        ent = -float(np.sum(probs * np.log2(probs + 1e-30)))
        entropies.append(ent)

    return entropies


# ─────────────────────────────────────────────────────────────────────
# Spectral Coherence Invariant (SCI) Tracker
# ─────────────────────────────────────────────────────────────────────

class SpectralCoherenceTracker:
    """
    Tracks the Spectral Coherence Invariant (SCI) over time.

    Verifies that C(S_{t+1}) ≥ C(S_t) · (1 − ε_t / ε_max)
    as guaranteed by the theorem.

    Also computes the Manifold Decay prediction for Betti-0 survival.
    """

    def __init__(self, window_size: int = 256, epsilon_max: float = 0.9):
        self.window_size = window_size
        self.epsilon_max = epsilon_max
        self._history: List[float] = []
        self._coherence_history: List[float] = []
        self._invariant_violations = 0
        self._total_checks = 0

    def update(self, delta: float, epsilon_t: float) -> Dict[str, float]:
        """
        Feed one I/O delta and current governor epsilon.

        Returns diagnostic dict with coherence, entropy, invariant status.
        """
        self._history.append(delta)

        # Keep rolling window
        if len(self._history) > self.window_size:
            self._history = self._history[-self.window_size:]

        if len(self._history) < 4:
            return {"coherence": 0.0, "entropy": 0.0, "invariant_holds": True}

        signal = np.array(self._history)
        C = spectral_coherence(signal)
        H = spectral_entropy(signal)

        # Check SCI
        invariant_holds = True
        if self._coherence_history:
            C_prev = self._coherence_history[-1]
            bound = C_prev * (1.0 - epsilon_t / self.epsilon_max)
            self._total_checks += 1
            if C < bound - 1e-6:  # Allow floating-point tolerance
                self._invariant_violations += 1
                invariant_holds = False

        self._coherence_history.append(C)

        return {
            "coherence": C,
            "spectral_entropy": H,
            "invariant_holds": invariant_holds,
            "violation_rate": self._invariant_violations / max(self._total_checks, 1),
            "window_size": len(self._history),
        }

    def predict_betti_survival(self, betti_0_initial: int, time_elapsed: float,
                               decay_lambda: float,
                               reinforcement_density: float) -> float:
        """
        Manifold Decay Theorem prediction.

        β₀(M_t) ≤ β₀(M₀) · exp(−λt · (1 − ρ))

        Parameters
        ----------
        betti_0_initial : Initial number of connected components
        time_elapsed : Time since manifold creation
        decay_lambda : Memory decay rate
        reinforcement_density : Fraction of reinforced memories ρ ∈ [0, 1]
        """
        rho = max(0.0, min(1.0, reinforcement_density))
        exponent = -decay_lambda * time_elapsed * (1.0 - rho)
        return betti_0_initial * math.exp(exponent)


# ─────────────────────────────────────────────────────────────────────
# Anomaly Detection via Spectral Divergence
# ─────────────────────────────────────────────────────────────────────

def spectral_kl_divergence(signal_a: np.ndarray, signal_b: np.ndarray) -> float:
    """
    KL divergence between the power spectra of two signals.

    D_KL(P_A || P_B) = Σ P_A(k) · log(P_A(k) / P_B(k))

    High divergence = the signals have very different frequency structure.
    Used to detect when I/O patterns shift (e.g., sequential → random).
    """
    def _power_dist(sig):
        fft = np.fft.rfft(sig)
        power = np.abs(fft[1:]) ** 2
        total = float(np.sum(power))
        if total < 1e-24:
            return np.ones(len(power)) / len(power)
        return power / total

    # Ensure same length
    min_len = min(len(signal_a), len(signal_b))
    p = _power_dist(signal_a[:min_len])
    q = _power_dist(signal_b[:min_len])

    # Symmetric KL with epsilon for stability
    eps = 1e-10
    kl = float(np.sum(p * np.log((p + eps) / (q + eps))))
    return max(0.0, kl)

File: kernel/epsilon/epsilon/crates/aether-core/src/spectral_entropy.rs
// Epsilon-Hollow - Copyright (c) 2024 Teerth Sharma
// SPDX-License-Identifier: Epsilon-Hollow

//! no_std spectral coherence, entropy, Haar wavelet entropy, and decay bounds.

use alloc::vec::Vec;

const EPS: f64 = 1e-24;
const LOG2: f64 = core::f64::consts::LN_2;

/// Compute spectral coherence from the non-DC DFT power spectrum.
///
/// This mirrors the Python formula `sum(|F[k]|^2) / (N * variance)`.
pub fn spectral_coherence(signal: &[f64]) -> f64 {
    let n = signal.len();
    if n < 2 {
        return 0.0;
    }

    let var = variance(signal);
    if var < EPS {
        return n as f64;
    }

    dft_non_dc_power(signal).into_iter().sum::<f64>() / (n as f64 * var)
}

/// Compute Shannon entropy of the normalized non-DC DFT power spectrum.
pub fn spectral_entropy(signal: &[f64]) -> f64 {
    let power = dft_non_dc_power(signal);
    entropy_from_energy(&power)
}

/// Coefficients returned by [`haar_wavelet_transform`].
#[derive(Debug, Clone, PartialEq)]
pub struct HaarWaveletTransform {
    /// Final approximation coefficients after all Haar levels.
    pub approximation: Vec<f64>,
    /// Detail coefficients at each level, finest to coarsest.
    pub details: Vec<Vec<f64>>,
}

/// Compute the discrete Haar wavelet transform.
///
/// The input is zero-padded to the next power of two. The returned details
/// match the legacy ordering: level 1 first, then progressively coarser levels.
pub fn haar_wavelet_transform(signal: &[f64]) -> HaarWaveletTransform {
    let n = signal.len();
    let mut power = 1usize;
    while power < n {
        power *= 2;
    }

    let mut current = Vec::with_capacity(power);
    current.resize(power, 0.0);
    let mut i = 0;
    while i < n {
        current[i] = signal[i];
        i += 1;
    }

    let mut details = Vec::new();
    while current.len() > 1 {
        let half = current.len() / 2;
        let mut approx = Vec::with_capacity(half);
        approx.resize(half, 0.0);
        let mut detail = Vec::with_capacity(half);
        detail.resize(half, 0.0);

        let mut j = 0;
        while j < half {
            let a = current[2 * j];
            let b = current[2 * j + 1];
            approx[j] = (a + b) / libm::sqrt(2.0);
            detail[j] = (a - b) / libm::sqrt(2.0);
            j += 1;
        }

        details.push(detail);
        current = approx;
    }

    HaarWaveletTransform {
        approximation: current,
        details,
    }
}

/// Compute Haar wavelet entropy for each scale, finest to coarsest.
pub fn wavelet_entropy_per_scale(signal: &[f64]) -> Vec<f64> {
    let transform = haar_wavelet_transform(signal);
    let mut entropies = Vec::new();

    for detail in transform.details {
        let mut energy = Vec::with_capacity(detail.len());
        energy.resize(detail.len(), 0.0);
        let mut i = 0;
        while i < detail.len() {
            energy[i] = detail[i] * detail[i];
            i += 1;
        }

        entropies.push(entropy_from_energy(&energy));
    }

    entropies
}

/// KL divergence between normalized non-DC power spectra.
pub fn spectral_kl_divergence(signal_a: &[f64], signal_b: &[f64]) -> f64 {
    let n = signal_a.len().min(signal_b.len());
    if n < 2 {
        return 0.0;
    }

    let p_power = dft_non_dc_power(&signal_a[..n]);
    let q_power = dft_non_dc_power(&signal_b[..n]);
    let p = normalized_distribution(&p_power);
    let q = normalized_distribution(&q_power);

    let mut kl = 0.0;
    let mut i = 0;
    while i < p.len().min(q.len()) {
        kl += p[i] * libm::log((p[i] + 1e-10) / (q[i] + 1e-10));
        i += 1;
    }
    kl.max(0.0)
}

/// Result returned by [`SpectralCoherenceTracker::update`].
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SpectralUpdate {
    /// Current spectral coherence.
    pub coherence: f64,
    /// Current spectral entropy.
    pub spectral_entropy: f64,
    /// Whether the spectral coherence invariant held for this update.
    pub invariant_holds: bool,
    /// Running invariant violation rate.
    pub violation_rate: f64,
    /// Number of samples currently retained in the rolling window.
    pub window_size: usize,
}

impl Default for SpectralUpdate {
    fn default() -> Self {
        Self {
            coherence: 0.0,
            spectral_entropy: 0.0,
            invariant_holds: true,
            violation_rate: 0.0,
            window_size: 0,
        }
    }
}

/// Rolling tracker for the Spectral Coherence Invariant theorem.
#[derive(Debug, Clone)]
pub struct SpectralCoherenceTracker {
    window_size: usize,
    epsilon_max: f64,
    history: Vec<f64>,
    coherence_history: Vec<f64>,
    invariant_violations: usize,
    total_checks: usize,
}

impl SpectralCoherenceTracker {
    /// Construct a tracker with a rolling window and maximum governor epsilon.
    pub fn new(window_size: usize, epsilon_max: f64) -> Self {
        Self {
            window_size: window_size.max(1),
            epsilon_max,
            history: Vec::new(),
            coherence_history: Vec::new(),
            invariant_violations: 0,
            total_checks: 0,
        }
    }

    /// Feed one I/O delta and check the spectral coherence invariant.
    pub fn update(&mut self, delta: f64, epsilon_t: f64) -> SpectralUpdate {
        self.history.push(delta);
        while self.history.len() > self.window_size {
            self.history.remove(0);
        }

        if self.history.len() < 4 {
            return SpectralUpdate {
                window_size: self.history.len(),
                ..SpectralUpdate::default()
            };
        }

        let coherence = spectral_coherence(&self.history);
        let spectral_entropy = spectral_entropy(&self.history);
        let mut invariant_holds = true;

        if let Some(&previous) = self.coherence_history.last() {
            let factor = if self.epsilon_max <= 0.0 {
                0.0
            } else {
                1.0 - epsilon_t / self.epsilon_max
            };
            let bound = previous * factor;
            self.total_checks += 1;
            if coherence < bound - 1e-6 {
                self.invariant_violations += 1;
                invariant_holds = false;
            }
        }
        self.coherence_history.push(coherence);

        SpectralUpdate {
            coherence,
            spectral_entropy,
            invariant_holds,
            violation_rate: self.violation_rate(),
            window_size: self.history.len(),
        }
    }

    /// Predict the Manifold Decay Theorem Betti-0 survival bound.
    pub fn predict_betti_survival(
        &self,
        betti_0_initial: usize,
        time_elapsed: f64,
        decay_lambda: f64,
        reinforcement_density: f64,
    ) -> f64 {
        let rho = reinforcement_density.clamp(0.0, 1.0);
        let exponent = -decay_lambda * time_elapsed * (1.0 - rho);
        betti_0_initial as f64 * libm::exp(exponent)
    }

    /// Return the running invariant violation rate.
    #[inline]
    pub fn violation_rate(&self) -> f64 {
        if self.total_checks == 0 {
            0.0
        } else {
            self.invariant_violations as f64 / self.total_checks as f64
        }
    }
}

fn variance(signal: &[f64]) -> f64 {
    let mean = signal.iter().copied().sum::<f64>() / signal.len() as f64;
    let mut acc = 0.0;
    for &x in signal {
        let d = x - mean;
        acc += d * d;
    }
    acc / signal.len() as f64
}

fn dft_non_dc_power(signal: &[f64]) -> Vec<f64> {
    let n = signal.len();
    if n < 2 {
        return Vec::new();
    }

    let bins = n / 2;
    let mut power = Vec::with_capacity(bins);
    let mut k = 1usize;
    while k <= bins {
        let mut re = 0.0;
        let mut im = 0.0;
        let mut t = 0usize;
        while t < n {
            let angle = -2.0 * core::f64::consts::PI * k as f64 * t as f64 / n as f64;
            re += signal[t] * libm::cos(angle);
            im += signal[t] * libm::sin(angle);
            t += 1;
        }
        power.push(re * re + im * im);
        k += 1;
    }
    power
}

fn entropy_from_energy(energy: &[f64]) -> f64 {
    let total = energy.iter().copied().sum::<f64>();
    if total < EPS {
        return 0.0;
    }

    let mut entropy = 0.0;
    for &e in energy {
        let p = e / total;
        if p > 0.0 {
            entropy -= p * (libm::log(p) / LOG2);
        }
    }
    entropy
}

fn normalized_distribution(power: &[f64]) -> Vec<f64> {
    let mut out = Vec::with_capacity(power.len());
    let total = power.iter().copied().sum::<f64>();
    if power.is_empty() {
        return out;
    }

    if total < EPS {
        let uniform = 1.0 / power.len() as f64;
        let mut i = 0;
        while i < power.len() {
            out.push(uniform);
            i += 1;
        }
    } else {
        for &p in power {
            out.push(p / total);
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn spectral_metrics_identify_periodic_and_constant_signals() {
        let periodic = [1.0, -1.0, 1.0, -1.0];
        assert!((spectral_coherence(&periodic) - 4.0).abs() < 1e-12);
        assert!(spectral_entropy(&periodic) < 1e-9);

        let constant = [3.0, 3.0, 3.0, 3.0];
        assert_eq!(spectral_coherence(&constant), 4.0);
        assert_eq!(spectral_entropy(&constant), 0.0);
    }

    #[test]
    fn haar_entropy_and_tracker_cover_sci_and_decay() {
        let signal = [1.0, 1.0, -1.0, -1.0];
        let transform = haar_wavelet_transform(&signal);
        assert_eq!(transform.approximation.len(), 1);
        assert!((transform.approximation[0] - 0.0).abs() < 1e-12);
        assert_eq!(transform.details.len(), 2);
        assert_eq!(transform.details[0].len(), 2);
        assert!((transform.details[1][0] - 2.0).abs() < 1e-12);

        let entropies = wavelet_entropy_per_scale(&signal);
        assert_eq!(entropies.len(), 2);
        assert!(entropies[0] <= 1.0);

        let mut tracker = SpectralCoherenceTracker::new(8, 0.9);
        let mut last = SpectralUpdate::default();
        for delta in [1.0, -1.0, 1.0, -1.0, 1.0] {
            last = tracker.update(delta, 0.1);
        }
        assert!(last.window_size >= 4);
        assert!(last.coherence > 0.0);
        assert!(last.violation_rate >= 0.0);

        let survival = tracker.predict_betti_survival(10, 2.0, 0.5, 0.5);
        assert!(survival > 6.0 && survival < 7.0);
    }

    #[test]
    fn spectral_kl_divergence_is_zero_for_identical_signals() {
        let signal = [0.0, 1.0, 0.0, -1.0, 0.0, 1.0, 0.0, -1.0];
        assert!(spectral_kl_divergence(&signal, &signal) < 1e-12);
    }
}
